ompi/mca/osc/ucx/osc_ucx_comm.c

```diff
@@ -427,19 +453,42 @@ int ompi_osc_ucx_get(void *origin_addr, int origin_count,
                      int target, ptrdiff_t target_disp, int target_count,
                      struct ompi_datatype_t *target_dt, struct ompi_win_t *win) {
     ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
-    ucp_ep_h ep = OSC_UCX_GET_EP(module->comm, target);
+    ucp_ep_h ep;
     uint64_t remote_addr = (module->win_info_array[target]).addr + target_disp * OSC_UCX_GET_DISP(module, target);
     ucp_rkey_h rkey;
     ptrdiff_t origin_lb, origin_extent, target_lb, target_extent;
     bool is_origin_contig = false, is_target_contig = false;
     ucs_status_t status;
+    pthread_t tid = pthread_self();
     int ret = OMPI_SUCCESS;
 
     ret = check_sync_state(module, target, false);
     if (ret != OMPI_SUCCESS) {
         return ret;
     }
 
+    if (pthread_equal(tid, mca_osc_ucx_component.main_tid)) {
+        ep = OSC_UCX_GET_EP(module->comm, target);
+        rkey = (module->win_info_array[target]).rkey;
+    } else {
+        thread_local_info_t *curr_thread_info;
+        if ((curr_thread_info = pthread_getspecific(my_thread_key)) == NULL) {
+            ret = opal_common_ucx_create_local_worker(mca_osc_ucx_component.ucp_context,
+                                                      ompi_comm_size(module->comm),
+                                                      mca_osc_ucx_component.worker_addr_buf,
+                                                      mca_osc_ucx_component.worker_addr_disps,
+                                                      mca_osc_ucx_component.mem_addr_buf,
+                                                      mca_osc_ucx_component.mem_addr_disps);
+            if (ret != OMPI_SUCCESS) {
+                return ret;
+            }
+        }
+
+        curr_thread_info = pthread_getspecific(my_thread_key);
```

Comment:
I would shift it in the "if" statement. 
Otherwise you are always calling pthread_getspecific twice which we don't want